runtime/interp.c

```diff
@@ -284,23 +284,25 @@ value caml_interprete(code_t prog, asize_t prog_size)
 #endif
 
   if (prog == NULL) {           /* Interpreter is initializing */
-    static opcode_t raise_unhandled_code[] = { ACC, 0, RAISE };
-    value raise_unhandled_closure;
+    static opcode_t raise_unhandled_effect_code[] = { ACC, 0, RAISE };
+    value raise_unhandled_effect_closure;
 
     caml_register_code_fragment(
-      (char *) raise_unhandled_code,
-      (char *) raise_unhandled_code + sizeof(raise_unhandled_code),
+      (char *) raise_unhandled_effect_code,
+      (char *) raise_unhandled_effect_code +
+      sizeof(raise_unhandled_effect_code),
       DIGEST_IGNORE, NULL);
 #ifdef THREADED_CODE
     caml_init_thread_code(jumptable, Jumptbl_base);
-    caml_thread_code(raise_unhandled_code,
-                     sizeof(raise_unhandled_code));
+    caml_thread_code(raise_unhandled_effect_code,
+                     sizeof(raise_unhandled_effect_code));
 #endif
-    raise_unhandled_closure = caml_alloc_small (2, Closure_tag);
-    Code_val(raise_unhandled_closure) = (code_t)raise_unhandled_code;
-    Closinfo_val(raise_unhandled_closure) = Make_closinfo(0, 2);
-    raise_unhandled = raise_unhandled_closure;
-    caml_register_generational_global_root(&raise_unhandled);
+    raise_unhandled_effect_closure = caml_alloc_small (2, Closure_tag);
+    Code_val(raise_unhandled_effect_closure) =
+      (code_t)raise_unhandled_effect_code;
+    Closinfo_val(raise_unhandled_effect_closure) = Make_closinfo(0, 2);
+    raise_unhandled_effect = raise_unhandled_effect_closure;
+    caml_register_generational_global_root(&raise_unhandled_effect);
```

Comment:
Can you explain to a newcomer why this black magic around `raise_unhandled_effect` is necessary? (The explanation could be added as a comment in the code.)

I'm not sure what parts of the bytecode are allowed to directly raise exceptions (apparently `caml_raise_continuation_already_taken()` below is fine, and there are existing examples with `caml_raise_stack_overflow` for example), and which are not / why. Here what's going on seems very complex compared to the semantics we want, or even to the handling of unhandled exceptions in

https://github.com/ocaml/ocaml/blob/37df9347658338f488e74780d9e1b0cf9eecc4fc/runtime/interp.c#L972-L977